File: dotnet/Watchup/Transport.cs
// ─────────────────────────────────────────────────────────────────────────────
// Watchup .NET SDK  ·  Transport
//
// Thin HTTP layer around HttpClient.
// All failures are swallowed — the SDK must never crash the host application.
// ─────────────────────────────────────────────────────────────────────────────

using System.Net.Http.Json;
using System.Text.Json;

namespace Watchup;

internal sealed class Transport
{
    private readonly HttpClient    _http;
    private readonly string        _url;
    private readonly bool          _debug;

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
    };

    internal Transport(HttpClient http, string baseUrl, bool debug)
    {
        _http  = http;
        _url   = $"{baseUrl.TrimEnd('/')}/api/v1/ingest/batch";
        _debug = debug;
    }

    /// <summary>
    /// POST <paramref name="batch"/> to the Watchup ingest endpoint.
    /// Never throws — any error is logged (if debug) and silently dropped.
    /// </summary>
    internal async Task SendAsync(IngestBatch batch, CancellationToken ct = default)
    {
        try
        {
            using var response = await _http
                .PostAsJsonAsync(_url, batch, JsonOpts, ct)
                .ConfigureAwait(false);

            if (_debug && !response.IsSuccessStatusCode)
            {
                var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                Console.Error.WriteLine($"[watchup] ingest {(int)response.StatusCode}: {body}");
            }
        }
        catch (Exception ex) when (_debug)
        {
            Console.Error.WriteLine($"[watchup] send failed: {ex.Message}");
        }
        catch
        {
            // Intentionally swallowed — SDK must never crash the host.
        }
    }
}

File: node/src/transport.ts
// ─────────────────────────────────────────────────────────────────────────────
// @watchupltd/node  ·  transport
//
// Thin HTTP layer around fetch (Node ≥ 18 built-in).
// All failures are swallowed — the SDK must never crash the host application.
// ─────────────────────────────────────────────────────────────────────────────

import type { IngestBatch } from './types.js';

export class Transport {
  private readonly url: string;
  private readonly headers: Record<string, string>;
  private readonly debug: boolean;

  constructor(baseUrl: string, apiKey: string, debug = false) {
    this.url   = `${baseUrl.replace(/\/$/, '')}/api/v1/ingest/batch`;
    this.headers = {
      'Content-Type': 'application/json',
      'X-Api-Key':    apiKey,
      'User-Agent':   `@watchupltd/node`,
    };
    this.debug = debug;
  }

  /**
   * POST `batch` to the Watchup ingest endpoint.
   * Never rejects — any error is logged (if debug) and silently dropped.
   */
  async send(batch: IngestBatch): Promise<void> {
    try {
      const res = await fetch(this.url, {
        method:  'POST',
        headers: this.headers,
        body:    JSON.stringify(batch),
        // 8-second hard timeout; slow API shouldn't hold up the queue.
        signal:  AbortSignal.timeout(8_000),
      });

      if (this.debug && !res.ok) {
        const body = await res.text().catch(() => '(no body)');
        console.warn(`[watchup] ingest ${res.status}: ${body}`);
      }
    } catch (err) {
      if (this.debug) {
        const msg = err instanceof Error ? err.message : String(err);
        console.warn(`[watchup] send failed: ${msg}`);
      }
      // Intentionally no re-throw.
    }
  }
}
